core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java

```diff
@@ -198,10 +198,19 @@ public StructuredRecord joinOn(String stageName, StructuredRecord record) {
     List<Schema.Field> fields = new ArrayList<>();
     Schema schema = record.getSchema();
 
+
     List<String> joinKeys = perStageJoinKeys.get(stageName);
     int i = 1;
     for (String joinKey : joinKeys) {
-      Schema.Field joinField = Schema.Field.of(String.valueOf(i++), schema.getField(joinKey).getSchema());
+      Schema.Field field = schema.getField(joinKey);
+      if (field == null) {
+        collector.addFailure(
```

Comment:
collector isn't really needed in this scenario, just throw an exception directly. The collector is useful mostly at configure time, to collect more than one failure and display it to the user at once.

Reply:
updated to throw IllegalArgumentException